### Capstone Project - The Battle of Neighborhoods

## Business Problem section

# Background
According to Bloomberg News, the London Housing Market is in a rut. It is now facing a number of different headwinds, including the prospect of higher taxes and a warning from the Bank of England that U.K. home values could fall as much as 30 percent in the event of a disorderly exit from the European Union. More specifically, four overlooked cracks suggest that the London market may be in worse shape than many realize: hidden price falls, record-low sales, homebuilder exodus and tax hikes addressing overseas buyers of homes in England and Wales.

# Business Problem
In this scenario, it is urgent to adopt machine learning tools in order to assist homebuyers clientele in London to make wise and effective decisions. As a result, the business problem we are currently posing is: how could we provide support to homebuyers clientele in to purchase a suitable real estate in London in this uncertain economic and financial scenario?

To solve this business problem, we are going to cluster London neighborhoods in order to recommend venues and the current average price of real estate where homebuyers can make a real estate investment. We will recommend profitable venues according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores.

# Data section
Data on London properties and the relative price paid data were extracted from the HM Land Registry (http://landregistry.data.gov.uk/). The following fields comprise the address data included in Price Paid Data: Postcode; PAON Primary Addressable Object Name. Typically the house number or name; SAON Secondary Addressable Object Name. If there is a sub-building, for example, the building is divided into flats, there will be a SAON; Street; Locality; Town/City; District; County.

To explore and target recommended locations across different venues according to the presence of amenities and essential facilities, we will access data through FourSquare API interface and arrange them as a dataframe for visualization. By merging data on London properties and the relative price paid data from the HM Land Registry and data on amenities and essential facilities surrounding such properties from FourSquare API interface, we will be able to recommend profitable real estate investments.

 # Methodology section
The Methodology section will describe the main components of our analysis and predication system. The Methodology section comprises four stages:

1. Collect Inspection Data
2. Explore and Understand Data
3. Data preparation and preprocessing 
4. Modeling

# 1. Collect Inspection Data
After importing the necessary libraries, we download the data from the HM Land Registry website as follows

In [1]:
import os # Operating System
import numpy as np
import pandas as pd
import datetime as dt # Datetime
import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0 --yes
import folium #import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    geopy-1.18.1               |             py_0          51 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          84 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.49-py_0

The following packages will be UPDATED:

  geopy              conda-forge/linux-64::geopy-1.11.0-py~ --> conda-forge/noarch::geopy-1.18.1-py_0



geopy-1.18.1         | 51 KB     | ##################################### | 100% 
geographiclib-1.49   | 32 KB     | ##################################### | 100% 
Preparing tran

In [2]:
#Read the data for examination (Source: http://landregistry.data.gov.uk/)
df_ppd = pd.read_csv("http://prod2.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com/pp-2018.csv")

we anayize the data before using it

In [3]:
df_ppd.head(5)

{68FEB20B-A513-38DA-E053-6C04A8C051AE}  295000  2018-02-20 00:00   LE2 5PF  \
0  {68FEB20B-A514-38DA-E053-6C04A8C051AE}  450000  2018-03-15 00:00   LE9 9FH   
1  {68FEB20B-A515-38DA-E053-6C04A8C051AE}  980000  2018-02-12 00:00  LE10 3AB   
2  {68FEB20B-A516-38DA-E053-6C04A8C051AE}  158000  2018-03-01 00:00   LE2 8FQ   
3  {68FEB20B-A518-38DA-E053-6C04A8C051AE}  240000  2018-02-12 00:00  LE11 2HN   
4  {68FEB20B-A519-38DA-E053-6C04A8C051AE}  150000  2018-02-09 00:00   LE2 8QB   

   D  N  F               90 Unnamed: 8    HIDCOTE ROAD       OADBY  \
0  D  N  F  MARKFIELD LODGE        NaN  MARKFIELD LANE  BOTCHESTON   
1  D  N  F     COTTAGE FARM        NaN        FOSSEWAY   SHARNFORD   
2  T  N  F               69        NaN      PERCY ROAD         NaN   
3  D  N  F              360        NaN       PARK ROAD         NaN   
4  S  N  F               13        NaN      ALTON ROAD         NaN   

      LEICESTER      OADBY AND WIGSTON  LEICESTERSHIRE  A A.1  
0     LEICESTER  HINCKLEY AND BOSWORTH  LEICESTERSHIRE  A   A  
1      HINCKLEY                  BLABY  LEICESTERSHIRE  A   A  
2     LEICESTER              LEICESTER       LEICESTER  A   A  
3  LOUGHBOROUGH              CHARNWOOD  LEICESTERSHIRE  A   A  
4     LEICESTER              LEICESTER       LEICESTER  A   A

In [4]:
df_ppd.shape

(971140, 16)

Our dataset consists of over 971140 rows and 16 columns. We will now prepare and preprocess data accordingly.

# 3. Data preparation and preprocessing
At this stage, we prepare our dataset for the modeling process, opting for the most suitable machine learning algorithm for our scope. Accordingly, we perform the following steps:

Rename the column names
Format the date column
Sort data by date of sale
Select data only for the city of London
Make a list of street names in London
Calculate the street-wise average price of the property
Read the street-wise coordinates into a data frame, eliminating recurring word London from individual names
Join the data to find the coordinates of locations which fit into client's budget
Plot recommended locations on London map along with current market prices

In [5]:
# Assign meaningful column names
df_ppd.columns = ['TUID', 'Price', 'Date_Transfer', 'Postcode', 'Prop_Type', 'Old_New', 'Duration', 'PAON', \
                  'SAON', 'Street', 'Locality', 'Town_City', 'District', 'County', 'PPD_Cat_Type', 'Record_Status']

In [9]:
# Format the date column
df_ppd['Date_Transfer'] = df_ppd['Date_Transfer'].apply(pd.to_datetime)

# Delete all obsolete transactions which were done before 2016
df_ppd.drop(df_ppd[df_ppd.Date_Transfer.dt.year < 2016].index, inplace=True)

# Sort by Date of Sale
df_ppd.sort_values(by=['Date_Transfer'],ascending=[False],inplace=True)

In [10]:
df_ppd_london = df_ppd.query("Town_City == 'LONDON'")

# Make a list of street names in LONDON
streets = df_ppd_london['Street'].unique().tolist()

In [11]:
df_grp_price = df_ppd_london.groupby(['Street'])['Price'].mean().reset_index()

# Give meaningful names to the columns
df_grp_price.columns = ['Street', 'Avg_Price']

In [12]:
#Input your Budget's Upper Limit and Lower Limit - Find the locations df_grp_price which fits your budget
df_affordable = df_grp_price.query("(Avg_Price >= 2200000) & (Avg_Price <= 2500000)")

In [13]:
# Display the dataframe
df_affordable

Street     Avg_Price
192            ALBION SQUARE  2.450000e+06
384              ANHALT ROAD  2.435000e+06
399          ANSDELL TERRACE  2.250000e+06
413          APPLEGARTH ROAD  2.400000e+06
685         AYLESTONE AVENUE  2.286667e+06
838          BARONSMEAD ROAD  2.375000e+06
961           BEAUCLERC ROAD  2.480000e+06
1082         BELVEDERE DRIVE  2.340000e+06
1195       BICKENHALL STREET  2.208500e+06
1233       BIRCHLANDS AVENUE  2.217000e+06
1530          BRAMPTON GROVE  2.456875e+06
1609       BRIARDALE GARDENS  2.397132e+06
1772                BROOKWAY  2.400000e+06
1885            BURBAGE ROAD  2.445000e+06
1949               BURY WALK  2.492500e+06
2030         CALEDONIAN ROAD  2.349182e+06
2035         CALLCOTT STREET  2.375000e+06
2096       CAMPDEN HILL ROAD  2.379653e+06
2103            CAMPION ROAD  2.461000e+06
2125           CANNING PLACE  2.425000e+06
2192           CARLISLE ROAD  2.200000e+06
2197         CARLTON GARDENS  2.483500e+06
2209           CARLYLE COURT  2.300000e+06
2371          CHALCOT SQUARE  2.286679e+06
2449            CHARLES LANE  2.414000e+06
2525        CHELSEA CRESCENT  2.495000e+06
2569     CHESTER CLOSE NORTH  2.450000e+06
2600            CHEYNE COURT  2.250000e+06
2769        CLARENDON STREET  2.250000e+06
2859        CLONCURRY STREET  2.388333e+06
...                      ...           ...
10679             ROYAL HILL  2.436250e+06
10733   RUSSELL GARDENS MEWS  2.300000e+06
10974         SETTLES STREET  2.487500e+06
11048         SHELDON AVENUE  2.349542e+06
11192     SKEFFINGTON STREET  2.300000e+06
11276          SOUTH END ROW  2.470000e+06
11351    SOUTHWOOD LAWN ROAD  2.350000e+06
11354         SOVEREIGN PARK  2.500000e+06
11566  ST MARGARETS CRESCENT  2.216500e+06
11601       ST OSWALDS PLACE  2.250000e+06
11621       ST PETERS SQUARE  2.468730e+06
11658       STAFFORD TERRACE  2.355000e+06
11992       SUTHERLAND PLACE  2.456000e+06
12054          SYDNEY STREET  2.240833e+06
12194            THAMES BANK  2.400000e+06
12255            THE HEXAGON  2.335000e+06
12533        TREDEGAR SQUARE  2.436666e+06
12595         TRINITY STREET  2.317500e+06
12744   UPPER HAMPSTEAD WALK  2.500000e+06
13001        WALPOLE GARDENS  2.303500e+06
13076         WARWICK SQUARE  2.432273e+06
13172            WELBECK WAY  2.267000e+06
13184      WELLESLEY TERRACE  2.410000e+06
13326     WESTMORELAND PLACE  2.300000e+06
13476         WILFRED STREET  2.410538e+06
13501     WILLOW BRIDGE ROAD  2.425000e+06
13520          WILSON STREET  2.257500e+06
13547        WINCHENDON ROAD  2.350000e+06
13584           WINGATE ROAD  2.206400e+06
13797        WRENTHAM AVENUE  2.232500e+06

[148 rows x 2 columns]

In [14]:
import pandas as pd
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim
from geopy.distance import vincenty
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [15]:
for index, item in df_affordable.iterrows():
    print(f"index: {index}")
    print(f"item: {item}")
    print(f"item.Street only: {item.Street}")

index: 192
item: Street       ALBION SQUARE
Avg_Price         2.45e+06
Name: 192, dtype: object
item.Street only: ALBION SQUARE
index: 384
item: Street       ANHALT ROAD
Avg_Price      2.435e+06
Name: 384, dtype: object
item.Street only: ANHALT ROAD
index: 399
item: Street       ANSDELL TERRACE
Avg_Price           2.25e+06
Name: 399, dtype: object
item.Street only: ANSDELL TERRACE
index: 413
item: Street       APPLEGARTH ROAD
Avg_Price            2.4e+06
Name: 413, dtype: object
item.Street only: APPLEGARTH ROAD
index: 685
item: Street       AYLESTONE AVENUE
Avg_Price         2.28667e+06
Name: 685, dtype: object
item.Street only: AYLESTONE AVENUE
index: 838
item: Street       BARONSMEAD ROAD
Avg_Price          2.375e+06
Name: 838, dtype: object
item.Street only: BARONSMEAD ROAD
index: 961
item: Street       BEAUCLERC ROAD
Avg_Price          2.48e+06
Name: 961, dtype: object
item.Street only: BEAUCLERC ROAD
index: 1082
item: Street       BELVEDERE DRIVE
Avg_Price           2.34e+06
Name

In [16]:
geolocator = Nominatim()

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  """Entry point for launching an IPython kernel.


In [17]:
df_affordable['city_coord'] = df_affordable['Street'].apply(geolocator.geocode).apply(lambda x: (x.latitude, x.longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [18]:
df_affordable

Street     Avg_Price                         city_coord
192            ALBION SQUARE  2.450000e+06   (-41.27375755, 173.289393239104)
384              ANHALT ROAD  2.435000e+06           (51.4803265, -0.1667607)
399          ANSDELL TERRACE  2.250000e+06           (51.4998899, -0.1891027)
413          APPLEGARTH ROAD  2.400000e+06           (53.7486539, -0.3266704)
685         AYLESTONE AVENUE  2.286667e+06           (51.5409157, -0.2178742)
838          BARONSMEAD ROAD  2.375000e+06            (51.4773147, -0.239457)
961           BEAUCLERC ROAD  2.480000e+06           (51.4995771, -0.2290331)
1082         BELVEDERE DRIVE  2.340000e+06            (34.966857, -81.992424)
1195       BICKENHALL STREET  2.208500e+06           (51.5211969, -0.1589341)
1233       BIRCHLANDS AVENUE  2.217000e+06           (51.4483941, -0.1604676)
1530          BRAMPTON GROVE  2.456875e+06           (51.5703648, -0.2833944)
1609       BRIARDALE GARDENS  2.397132e+06           (51.5601748, -0.1954305)
1772                BROOKWAY  2.400000e+06    (45.4321849, -122.802811661158)
1885            BURBAGE ROAD  2.445000e+06           (51.4509669, -0.0952879)
1949               BURY WALK  2.492500e+06           (52.1455294, -0.4235933)
2030         CALEDONIAN ROAD  2.349182e+06            (51.548482, -0.1183585)
2035         CALLCOTT STREET  2.375000e+06           (51.5083499, -0.1983276)
2096       CAMPDEN HILL ROAD  2.379653e+06           (51.5064605, -0.1988955)
2103            CAMPION ROAD  2.461000e+06            (52.6813505, 0.9654391)
2125           CANNING PLACE  2.425000e+06           (51.4995696, -0.1842477)
2192           CARLISLE ROAD  2.200000e+06          (42.5490988, -71.4166686)
2197         CARLTON GARDENS  2.483500e+06   (-37.80194335, 144.971970171017)
2209           CARLYLE COURT  2.300000e+06           (32.6264366, -83.717601)
2371          CHALCOT SQUARE  2.286679e+06           (51.5414545, -0.1552649)
2449            CHARLES LANE  2.414000e+06           (40.2375254, -83.349229)
2525        CHELSEA CRESCENT  2.495000e+06            (34.524738, -85.448539)
2569     CHESTER CLOSE NORTH  2.450000e+06           (51.5292054, -0.1450813)
2600            CHEYNE COURT  2.250000e+06             (51.599677, 0.5256231)
2769        CLARENDON STREET  2.250000e+06         (-19.2743572, 146.8004036)
2859        CLONCURRY STREET  2.388333e+06           (51.4737632, -0.2162442)
...                      ...           ...                                ...
10679             ROYAL HILL  2.436250e+06           (50.5370393, -3.9525672)
10733   RUSSELL GARDENS MEWS  2.300000e+06           (51.4998689, -0.2118904)
10974         SETTLES STREET  2.487500e+06           (51.5153237, -0.0643266)
11048         SHELDON AVENUE  2.349542e+06            (43.599287, -73.045426)
11192     SKEFFINGTON STREET  2.300000e+06           (52.3420958, -6.4632241)
11276          SOUTH END ROW  2.470000e+06           (51.4987463, -0.1890787)
11351    SOUTHWOOD LAWN ROAD  2.350000e+06            (51.5746267, -0.146238)
11354         SOVEREIGN PARK  2.500000e+06    (-27.4937089, 153.210145350143)
11566  ST MARGARETS CRESCENT  2.216500e+06            (52.2089792, 1.5662263)
11601       ST OSWALDS PLACE  2.250000e+06           (51.4872071, -0.1185341)
11621       ST PETERS SQUARE  2.468730e+06     (41.9022353, 12.4578392797771)
11658       STAFFORD TERRACE  2.355000e+06           (51.5009379, -0.1960492)
11992       SUTHERLAND PLACE  2.456000e+06            (51.6021232, 0.5230332)
12054          SYDNEY STREET  2.240833e+06           (-43.551026, 172.620521)
12194            THAMES BANK  2.400000e+06            (51.471715, -0.2689072)
12255            THE HEXAGON  2.335000e+06          (-31.8956726, 26.8758122)
12533        TREDEGAR SQUARE  2.436666e+06  (51.5270935, -0.0321842332596606)
12595         TRINITY STREET  2.317500e+06             (52.206141, 0.1181326)
12744   UPPER HAMPSTEAD WALK  2.500000e+06            (51.558467, -0.1774529)
13001        WALPOLE GARDENS  2.303500e+06  

In [19]:
df_affordable[['Latitude', 'Longitude']] = df_affordable['city_coord'].apply(pd.Series)

/home/jupyterlab/conda/lib/python3.6/site-packages/pandas/core/frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


In [20]:
df_affordable

Street     Avg_Price                         city_coord  \
192            ALBION SQUARE  2.450000e+06   (-41.27375755, 173.289393239104)   
384              ANHALT ROAD  2.435000e+06           (51.4803265, -0.1667607)   
399          ANSDELL TERRACE  2.250000e+06           (51.4998899, -0.1891027)   
413          APPLEGARTH ROAD  2.400000e+06           (53.7486539, -0.3266704)   
685         AYLESTONE AVENUE  2.286667e+06           (51.5409157, -0.2178742)   
838          BARONSMEAD ROAD  2.375000e+06            (51.4773147, -0.239457)   
961           BEAUCLERC ROAD  2.480000e+06           (51.4995771, -0.2290331)   
1082         BELVEDERE DRIVE  2.340000e+06            (34.966857, -81.992424)   
1195       BICKENHALL STREET  2.208500e+06           (51.5211969, -0.1589341)   
1233       BIRCHLANDS AVENUE  2.217000e+06           (51.4483941, -0.1604676)   
1530          BRAMPTON GROVE  2.456875e+06           (51.5703648, -0.2833944)   
1609       BRIARDALE GARDENS  2.397132e+06           (51.5601748, -0.1954305)   
1772                BROOKWAY  2.400000e+06    (45.4321849, -122.802811661158)   
1885            BURBAGE ROAD  2.445000e+06           (51.4509669, -0.0952879)   
1949               BURY WALK  2.492500e+06           (52.1455294, -0.4235933)   
2030         CALEDONIAN ROAD  2.349182e+06            (51.548482, -0.1183585)   
2035         CALLCOTT STREET  2.375000e+06           (51.5083499, -0.1983276)   
2096       CAMPDEN HILL ROAD  2.379653e+06           (51.5064605, -0.1988955)   
2103            CAMPION ROAD  2.461000e+06            (52.6813505, 0.9654391)   
2125           CANNING PLACE  2.425000e+06           (51.4995696, -0.1842477)   
2192           CARLISLE ROAD  2.200000e+06          (42.5490988, -71.4166686)   
2197         CARLTON GARDENS  2.483500e+06   (-37.80194335, 144.971970171017)   
2209           CARLYLE COURT  2.300000e+06           (32.6264366, -83.717601)   
2371          CHALCOT SQUARE  2.286679e+06           (51.5414545, -0.1552649)   
2449            CHARLES LANE  2.414000e+06           (40.2375254, -83.349229)   
2525        CHELSEA CRESCENT  2.495000e+06            (34.524738, -85.448539)   
2569     CHESTER CLOSE NORTH  2.450000e+06           (51.5292054, -0.1450813)   
2600            CHEYNE COURT  2.250000e+06             (51.599677, 0.5256231)   
2769        CLARENDON STREET  2.250000e+06         (-19.2743572, 146.8004036)   
2859        CLONCURRY STREET  2.388333e+06           (51.4737632, -0.2162442)   
...                      ...           ...                                ...   
10679             ROYAL HILL  2.436250e+06           (50.5370393, -3.9525672)   
10733   RUSSELL GARDENS MEWS  2.300000e+06           (51.4998689, -0.2118904)   
10974         SETTLES STREET  2.487500e+06           (51.5153237, -0.0643266)   
11048         SHELDON AVENUE  2.349542e+06            (43.599287, -73.045426)   
11192     SKEFFINGTON STREET  2.300000e+06           (52.3420958, -6.4632241)   
11276          SOUTH END ROW  2.470000e+06           (51.4987463, -0.1890787)   
11351    SOUTHWOOD LAWN ROAD  2.350000e+06            (51.5746267, -0.146238)   
11354         SOVEREIGN PARK  2.500000e+06    (-27.4937089, 153.210145350143)   
11566  ST MARGARETS CRESCENT  2.216500e+06            (52.2089792, 1.5662263)   
11601       ST OSWALDS PLACE  2.250000e+06           (51.4872071, -0.1185341)   
11621       ST PETERS SQUARE  2.468730e+06     (41.9022353, 12.4578392797771)   
11658       STAFFORD TERRACE  2.355000e+06           (51.5009379, -0.1960492)   
11992       SUTHERLAND PLACE  2.456000e+06            (51.6021232, 0.5230332)   
12054          SYDNEY STREET  2.240833e+06           (-43.551026, 172.620521)   
12194            THAMES BANK  2.400000e+06            (51.471715, -0.2689072)   
12255            THE HEXAGON  2.335000e+06          (-31.8956726, 26.8758122)   
12533        TREDEGAR SQUARE  2.436666e+06  (51.5270935, -0.0321842332596606)   
12595         TRINITY STREET  2.317500e+06           

In [21]:
df = df_affordable.drop(columns=['city_coord'])

In [22]:
df

Street     Avg_Price   Latitude   Longitude
192            ALBION SQUARE  2.450000e+06 -41.273758  173.289393
384              ANHALT ROAD  2.435000e+06  51.480326   -0.166761
399          ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103
413          APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670
685         AYLESTONE AVENUE  2.286667e+06  51.540916   -0.217874
838          BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457
961           BEAUCLERC ROAD  2.480000e+06  51.499577   -0.229033
1082         BELVEDERE DRIVE  2.340000e+06  34.966857  -81.992424
1195       BICKENHALL STREET  2.208500e+06  51.521197   -0.158934
1233       BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468
1530          BRAMPTON GROVE  2.456875e+06  51.570365   -0.283394
1609       BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431
1772                BROOKWAY  2.400000e+06  45.432185 -122.802812
1885            BURBAGE ROAD  2.445000e+06  51.450967   -0.095288
1949               BURY WALK  2.492500e+06  52.145529   -0.423593
2030         CALEDONIAN ROAD  2.349182e+06  51.548482   -0.118359
2035         CALLCOTT STREET  2.375000e+06  51.508350   -0.198328
2096       CAMPDEN HILL ROAD  2.379653e+06  51.506461   -0.198896
2103            CAMPION ROAD  2.461000e+06  52.681351    0.965439
2125           CANNING PLACE  2.425000e+06  51.499570   -0.184248
2192           CARLISLE ROAD  2.200000e+06  42.549099  -71.416669
2197         CARLTON GARDENS  2.483500e+06 -37.801943  144.971970
2209           CARLYLE COURT  2.300000e+06  32.626437  -83.717601
2371          CHALCOT SQUARE  2.286679e+06  51.541455   -0.155265
2449            CHARLES LANE  2.414000e+06  40.237525  -83.349229
2525        CHELSEA CRESCENT  2.495000e+06  34.524738  -85.448539
2569     CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081
2600            CHEYNE COURT  2.250000e+06  51.599677    0.525623
2769        CLARENDON STREET  2.250000e+06 -19.274357  146.800404
2859        CLONCURRY STREET  2.388333e+06  51.473763   -0.216244
...                      ...           ...        ...         ...
10679             ROYAL HILL  2.436250e+06  50.537039   -3.952567
10733   RUSSELL GARDENS MEWS  2.300000e+06  51.499869   -0.211890
10974         SETTLES STREET  2.487500e+06  51.515324   -0.064327
11048         SHELDON AVENUE  2.349542e+06  43.599287  -73.045426
11192     SKEFFINGTON STREET  2.300000e+06  52.342096   -6.463224
11276          SOUTH END ROW  2.470000e+06  51.498746   -0.189079
11351    SOUTHWOOD LAWN ROAD  2.350000e+06  51.574627   -0.146238
11354         SOVEREIGN PARK  2.500000e+06 -27.493709  153.210145
11566  ST MARGARETS CRESCENT  2.216500e+06  52.208979    1.566226
11601       ST OSWALDS PLACE  2.250000e+06  51.487207   -0.118534
11621       ST PETERS SQUARE  2.468730e+06  41.902235   12.457839
11658       STAFFORD TERRACE  2.355000e+06  51.500938   -0.196049
11992       SUTHERLAND PLACE  2.456000e+06  51.602123    0.523033
12054          SYDNEY STREET  2.240833e+06 -43.551026  172.620521
12194            THAMES BANK  2.400000e+06  51.471715   -0.268907
12255            THE HEXAGON  2.335000e+06 -31.895673   26.875812
12533        TREDEGAR SQUARE  2.436666e+06  51.527093   -0.032184
12595         TRINITY STREET  2.317500e+06  52.206141    0.118133
12744   UPPER HAMPSTEAD WALK  2.500000e+06  51.558467   -0.177453
13001        WALPOLE GARDENS  2.303500e+06  51.439743   -0.341263
13076         WARWICK SQUARE  2.432273e+06  33.738563 -117.846506
13172            WELBECK WAY  2.267000e+06  51.517119   -0.148458
13184      WELLESLEY TERRACE  2.410000e+06  51.529214   -0.092933
13326     WESTMORELAND PLACE  2.300000e+06  40.740224 -111.846323
13476         WILFRED STREET  2.410538e+06  42.415553  -82.981523
13501     WILLOW BRIDGE ROAD  2.425000e+06  51.543109   -0.095508
13520          WILSON STREET  2.257500e+06 -19.268054  146.804289
13547        WINCHENDON ROAD  2.350000e+06  51.476958   -0.203903
13584           WINGATE ROAD  2.206400e+06  27.456603  -82.137

In [23]:
address = 'London, UK'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London City are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of London City are 51.5073219, -0.1276474.


In [24]:
# create map of London using latitude and longitude values
map_london = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, price, street in zip(df['Latitude'], df['Longitude'], df['Avg_Price'], df['Street']):
    label = '{}, {}'.format(street, price)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  
    
map_london

In [25]:
#Define Foursquare Credentials and Version

CLIENT_ID = 'VP0505UDJZH40US0YKW5YPEQNY2GD5EXGGTZWJQPEHKRJGN2' # Foursquare ID
CLIENT_SECRET = 'D1ZABF3LVHG2GDFQ14SIRYZ24UQB1WDEVSHBR1EKFL55SAOB' # Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VP0505UDJZH40US0YKW5YPEQNY2GD5EXGGTZWJQPEHKRJGN2
CLIENT_SECRET:D1ZABF3LVHG2GDFQ14SIRYZ24UQB1WDEVSHBR1EKFL55SAOB


We can now proceed to the Modeling phase. We will analyze neighborhoods to recommend real estates where home buyers can make a real estate investment. We will then recommend profitable venues according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores.

# 4. Modeling
After exploring the dataset and gaining insights into it, we are ready to use the clustering methodology to analyze real estates. We will use the k-means clustering technique as it is fast and efficient in terms of computational cost, is highly flexible to account for mutations in real estate market in London and is accurate.

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Street', 
                  'Street Latitude', 
                  'Street Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [27]:
# Run the above function on each location and create a new dataframe called location_venues and display it.
location_venues = getNearbyVenues(names=df['Street'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

ALBION SQUARE
ANHALT ROAD
ANSDELL TERRACE
APPLEGARTH ROAD
AYLESTONE AVENUE
BARONSMEAD ROAD
BEAUCLERC ROAD
BELVEDERE DRIVE
BICKENHALL STREET
BIRCHLANDS AVENUE
BRAMPTON GROVE
BRIARDALE GARDENS
BROOKWAY
BURBAGE ROAD
BURY WALK
CALEDONIAN ROAD
CALLCOTT STREET
CAMPDEN HILL ROAD
CAMPION ROAD
CANNING PLACE
CARLISLE ROAD
CARLTON GARDENS
CARLYLE COURT
CHALCOT SQUARE
CHARLES LANE
CHELSEA CRESCENT
CHESTER CLOSE NORTH
CHEYNE COURT
CLARENDON STREET
CLONCURRY STREET
COLBECK MEWS
CORNWALL TERRACE MEWS
COURT LANE GARDENS
CRAVEN STREET
DALEBURY ROAD
DEWHURST ROAD
DORIA ROAD
DOWNSHIRE HILL
DUCHESS WALK
ECCLESTON SQUARE MEWS
EGBERT STREET
EGERTON PLACE
ELM PARK ROAD
FARM PLACE
FRANK DIXON WAY
FULTON MEWS
GERARD ROAD
GERRARD ROAD
GIBSON SQUARE
GIRDLERS ROAD
GLOUCESTER CRESCENT
GORDON PLACE
GRAFTON SQUARE
GRAHAM TERRACE
HARMAN DRIVE
HAVANNAH STREET
HAZLEWELL ROAD
HEREFORD MEWS
HERONDALE AVENUE
HIGHGATE HIGH STREET
HILLGATE PLACE
HOLLYCROFT AVENUE
HOLLYWOOD MEWS
HONEYWELL ROAD
HOXTON SQUARE
HUNTER ROAD
JACKS

In [28]:
location_venues

Street  Street Latitude  Street Longitude  \
0       ALBION SQUARE       -41.273758        173.289393   
1       ALBION SQUARE       -41.273758        173.289393   
2       ALBION SQUARE       -41.273758        173.289393   
3       ALBION SQUARE       -41.273758        173.289393   
4       ALBION SQUARE       -41.273758        173.289393   
5       ALBION SQUARE       -41.273758        173.289393   
6       ALBION SQUARE       -41.273758        173.289393   
7       ALBION SQUARE       -41.273758        173.289393   
8       ALBION SQUARE       -41.273758        173.289393   
9       ALBION SQUARE       -41.273758        173.289393   
10      ALBION SQUARE       -41.273758        173.289393   
11      ALBION SQUARE       -41.273758        173.289393   
12      ALBION SQUARE       -41.273758        173.289393   
13      ALBION SQUARE       -41.273758        173.289393   
14      ALBION SQUARE       -41.273758        173.289393   
15      ALBION SQUARE       -41.273758        173.289393   
16      ALBION SQUARE       -41.273758        173.289393   
17      ALBION SQUARE       -41.273758        173.289393   
18      ALBION SQUARE       -41.273758        173.289393   
19      ALBION SQUARE       -41.273758        173.289393   
20      ALBION SQUARE       -41.273758        173.289393   
21      ALBION SQUARE       -41.273758        173.289393   
22      ALBION SQUARE       -41.273758        173.289393   
23      ALBION SQUARE       -41.273758        173.289393   
24      ALBION SQUARE       -41.273758        173.289393   
25      ALBION SQUARE       -41.273758        173.289393   
26      ALBION SQUARE       -41.273758        173.289393   
27      ALBION SQUARE       -41.273758        173.289393   
28        ANHALT ROAD        51.480326         -0.166761   
29        ANHALT ROAD        51.480326         -0.166761   
...               ...              ...               ...   
5304  WINCHENDON ROAD        51.476958         -0.203903   
5305  WINCHENDON ROAD        51.476958         -0.203903   
5306  WINCHENDON ROAD        51.476958         -0.203903   
5307  WINCHENDON ROAD        51.476958         -0.203903   
5308  WINCHENDON ROAD        51.476958         -0.203903   
5309  WINCHENDON ROAD        51.476958         -0.203903   
5310  WINCHENDON ROAD        51.476958         -0.203903   
5311  WINCHENDON ROAD        51.476958         -0.203903   
5312  WINCHENDON ROAD        51.476958         -0.203903   
5313  WINCHENDON ROAD        51.476958         -0.203903   
5314  WINCHENDON ROAD        51.476958         -0.203903   
5315  WINCHENDON ROAD        51.476958         -0.203903   
5316  WINCHENDON ROAD        51.476958         -0.203903   
5317  WINCHENDON ROAD        51.476958         -0.203903   
5318  WINCHENDON ROAD        51.476958         -0.203903   
5319  WINCHENDON ROAD        51.476958         -0.203903   
5320  WINCHENDON ROAD        51.476958         -0.203903   
5321  WINCHENDON ROAD        51.476958         -0.203903   
5322  WINCHENDON ROAD        51.476958         -0.203903   
5323  WINCHENDON ROAD        51.476958         -0.203903   
5324  WINCHENDON ROAD        51.476958         -0.203903   
5325  WINCHENDON ROAD        51.476958         -0.203903   
5326  WINCHENDON ROAD        51.476958         -0.203903   
5327  WINCHENDON ROAD        51.476958         -0.203903   
5328  WINCHENDON ROAD        51.476958         -0.203903   
5329  WINCHENDON ROAD        51.476958         -0.203903   
5330  WINCHENDON ROAD        51.476958         -0.203903   
5331  WINCHENDON ROAD        51.476958         -0.203903   
5332  WINCHENDON ROAD        51.476958         -0.203903   
5333  WRENTHAM AVENUE        51.359758          1.094681   

                              Venue  Venue Latitude  Venue Longitude  \
0                    The Free House      -41.273340       173.287364   
1                   The Indian Cafe      -41.273308       173.286530   
2                   Queen's Gardens      -41.273671       173.291383   
3     

In [29]:
location_venues.groupby('Street').count()

Street Latitude  Street Longitude  Venue  \
Street                                                            
ALBION SQUARE                       28                28     28   
ANHALT ROAD                         15                15     15   
ANSDELL TERRACE                     55                55     55   
APPLEGARTH ROAD                      4                 4      4   
AYLESTONE AVENUE                     4                 4      4   
BARONSMEAD ROAD                     15                15     15   
BEAUCLERC ROAD                      29                29     29   
BELVEDERE DRIVE                      2                 2      2   
BICKENHALL STREET                   95                95     95   
BIRCHLANDS AVENUE                   10                10     10   
BRAMPTON GROVE                       3                 3      3   
BRIARDALE GARDENS                    5                 5      5   
BROOKWAY                             1                 1      1   
BURBAGE ROAD                        18                18     18   
BURY WALK                           11                11     11   
CALEDONIAN ROAD                     25                25     25   
CALLCOTT STREET                     56                56     56   
CAMPDEN HILL ROAD                   57                57     57   
CAMPION ROAD                         4                 4      4   
CANNING PLACE                       32                32     32   
CARLISLE ROAD                        3                 3      3   
CARLTON GARDENS                     62                62     62   
CHALCOT SQUARE                      58                58     58   
CHARLES LANE                        19                19     19   
CHESTER CLOSE NORTH                 18                18     18   
CHEYNE COURT                         2                 2      2   
CLARENDON STREET                     4                 4      4   
CLONCURRY STREET                    27                27     27   
COLBECK MEWS                       100               100    100   
CORNWALL TERRACE MEWS               69                69     69   
...                                ...               ...    ...   
RODERICK ROAD                       18                18     18   
ROPEMAKERS FIELDS                   26                26     26   
ROYAL CRESCENT                      33                33     33   
RUSSELL GARDENS MEWS                26                26     26   
SETTLES STREET                      71                71     71   
SHELDON AVENUE                       3                 3      3   
SKEFFINGTON STREET                   6                 6      6   
SOUTH END ROW                       61                61     61   
SOUTHWOOD LAWN ROAD                 38                38     38   
SOVEREIGN PARK                       4                 4      4   
ST OSWALDS PLACE                    40                40     40   
ST PETERS SQUARE                    93                93     93   
STAFFORD TERRACE                    68                68     68   
SUTHERLAND PLACE                     2                 2      2   
SYDNEY STREET                        2                 2      2   
THAMES BANK                         15                15     15   
THE HEXAGON                          6                 6      6   
TREDEGAR SQUARE                     26                26     26   
TRINITY STREET                      90                90     90   
UPPER HAMPSTEAD WALK                57                57     57   
WALPOLE GARDENS                      8                 8      8   
WARWICK SQUARE                       4                 4      4   
WELBECK WAY                        100               100    100   
WELLESLEY TERRACE                   52                52     52   
WESTMORELAND PLACE                  16                16     16   
WILFRED STREET                       2                 2      2   
WILLOW BRIDGE ROAD                  25                25     25   
WIL

In [30]:
# get the List of Unique Categories
print('There are {} uniques categories.'.format(len(location_venues['Venue Category'].unique())))

There are 335 uniques categories.


In [31]:
# one hot encoding
venues_onehot = pd.get_dummies(location_venues[['Venue Category']], prefix="", prefix_sep="")

# add street column back to dataframe
venues_onehot['Street'] = location_venues['Street'] 

# move street column to the first column
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])

#fixed_columns
venues_onehot = venues_onehot[fixed_columns]

venues_onehot.head()

Street  Accessories Store  Afghan Restaurant  African Restaurant  \
0  ALBION SQUARE                  0                  0                   0   
1  ALBION SQUARE                  0                  0                   0   
2  ALBION SQUARE                  0                  0                   0   
3  ALBION SQUARE                  0                  0                   0   
4  ALBION SQUARE                  0                  0                   0   

   American Restaurant  Antique Shop  Arcade  Argentinian Restaurant  \
0                    0             0       0                       0   
1                    0             0       0                       0   
2                    0             0       0                       0   
3                    0             0       0                       0   
4                    0             0       0                       0   

   Art Gallery  Art Museum  ...  Video Game Store  Video Store  \
0            0           0  ...                 0            0   
1            0           0  ...                 0            0   
2            0           0  ...                 0            0   
3            0           0  ...                 0            0   
4            0           0  ...                 0            0   

   Vietnamese Restaurant  Warehouse Store  Waterfront  Windmill  Wine Bar  \
0                      0                0           0         0         0   
1                      0                0           0         0         0   
2                      0                0           0         0         0   
3                      0                0           0         0         0   
4                      0                0           0         0         0   

   Wine Shop  Women's Store  Yoga Studio  
0          0              0            0  
1          0              0            0  
2          0              0            0  
3          0              0            0  
4          0              0            0  

[5 rows x 336 columns]

In [32]:
london_grouped = venues_onehot.groupby('Street').mean().reset_index()
london_grouped

Street  Accessories Store  Afghan Restaurant  \
0            ALBION SQUARE                0.0                0.0   
1              ANHALT ROAD                0.0                0.0   
2          ANSDELL TERRACE                0.0                0.0   
3          APPLEGARTH ROAD                0.0                0.0   
4         AYLESTONE AVENUE                0.0                0.0   
5          BARONSMEAD ROAD                0.0                0.0   
6           BEAUCLERC ROAD                0.0                0.0   
7          BELVEDERE DRIVE                0.0                0.0   
8        BICKENHALL STREET                0.0                0.0   
9        BIRCHLANDS AVENUE                0.0                0.0   
10          BRAMPTON GROVE                0.0                0.0   
11       BRIARDALE GARDENS                0.0                0.0   
12                BROOKWAY                0.0                0.0   
13            BURBAGE ROAD                0.0                0.0   
14               BURY WALK                0.0                0.0   
15         CALEDONIAN ROAD                0.0                0.0   
16         CALLCOTT STREET                0.0                0.0   
17       CAMPDEN HILL ROAD                0.0                0.0   
18            CAMPION ROAD                0.0                0.0   
19           CANNING PLACE                0.0                0.0   
20           CARLISLE ROAD                0.0                0.0   
21         CARLTON GARDENS                0.0                0.0   
22          CHALCOT SQUARE                0.0                0.0   
23            CHARLES LANE                0.0                0.0   
24     CHESTER CLOSE NORTH                0.0                0.0   
25            CHEYNE COURT                0.0                0.0   
26        CLARENDON STREET                0.0                0.0   
27        CLONCURRY STREET                0.0                0.0   
28            COLBECK MEWS                0.0                0.0   
29   CORNWALL TERRACE MEWS                0.0                0.0   
..                     ...                ...                ...   
107          RODERICK ROAD                0.0                0.0   
108      ROPEMAKERS FIELDS                0.0                0.0   
109         ROYAL CRESCENT                0.0                0.0   
110   RUSSELL GARDENS MEWS                0.0                0.0   
111         SETTLES STREET                0.0                0.0   
112         SHELDON AVENUE                0.0                0.0   
113     SKEFFINGTON STREET                0.0                0.0   
114          SOUTH END ROW                0.0                0.0   
115    SOUTHWOOD LAWN ROAD                0.0                0.0   
116         SOVEREIGN PARK                0.0                0.0   
117       ST OSWALDS PLACE                0.0                0.0   
118       ST PETERS SQUARE                0.0                0.0   
119       STAFFORD TERRACE                0.0                0.0   
120       SUTHERLAND PLACE                0.0                0.0   
121          SYDNEY STREET                0.0                0.0   
122            THAMES BANK                0.0                0.0   
123            THE HEXAGON                0.0                0.0   
124        TREDEGAR SQUARE                0.0                0.0   
125         TRINITY STREET                0.0                0.0   
126   UPPER HAMPSTEAD WALK                0.0                0.0   
127        WALPOLE GARDENS                0.0                0.0   
128         WARWICK SQUARE                0.0                0.0   
129            WELBECK WAY                0.0                0.0   
130      WELLESLEY TERRACE                0.0                0.0   
131     WESTMORELAND PLACE                0.0                0.0   
132         WILFRED STREET                0.0                0.0   
133     WILLOW BRIDGE ROAD                0.0                0.0   
134     

In [33]:
london_grouped.shape

(137, 336)

In [34]:
# What are the top 5 venues/facilities nearby profitable real estate investments?#

num_top_venues = 5

for hood in london_grouped['Street']:
    print("----"+hood+"----")
    temp = london_grouped[london_grouped['Street'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----ALBION SQUARE----
               venue  freq
0               Café  0.21
1  Indian Restaurant  0.11
2                Bar  0.07
3                Pub  0.07
4         Restaurant  0.07


----ANHALT ROAD----
           venue  freq
0            Pub  0.27
1  Grocery Store  0.13
2          Plaza  0.07
3         Garden  0.07
4   Cocktail Bar  0.07


----ANSDELL TERRACE----
                venue  freq
0  Italian Restaurant  0.07
1      Clothing Store  0.07
2                 Pub  0.05
3               Hotel  0.05
4          Restaurant  0.04


----APPLEGARTH ROAD----
               venue  freq
0          Nightclub  0.25
1                Pub  0.25
2                Bar  0.25
3             Casino  0.25
4  Accessories Store  0.00


----AYLESTONE AVENUE----
               venue  freq
0               Park  0.50
1           Bus Stop  0.25
2      Movie Theater  0.25
3  Accessories Store  0.00
4  Outdoor Sculpture  0.00


----BARONSMEAD ROAD----
              venue  freq
0        Restaurant  0.13
1      

In [35]:
# Define a function to return the most common venues/facilities nearby real estate investments#

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Street']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [37]:
# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Street'] = london_grouped['Street']

for ind in np.arange(london_grouped.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(london_grouped.iloc[ind, :], num_top_venues)

In [38]:
venues_sorted.head()

Street 1st Most Common Venue 2nd Most Common Venue  \
0     ALBION SQUARE                  Café     Indian Restaurant   
1       ANHALT ROAD                   Pub         Grocery Store   
2   ANSDELL TERRACE    Italian Restaurant        Clothing Store   
3   APPLEGARTH ROAD                   Pub                   Bar   
4  AYLESTONE AVENUE                  Park         Movie Theater   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub            Restaurant           Coffee Shop   
1                Garden    Chinese Restaurant    English Restaurant   
2                 Hotel                   Pub     Indian Restaurant   
3             Nightclub                Casino           Fish Market   
4              Bus Stop           Flea Market    Falafel Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Bar                Museum          Burger Joint   
1          Cocktail Bar                 Diner           Pizza Place   
2                Bakery                Garden             Juice Bar   
3    Falafel Restaurant                  Farm        Farmers Market   
4                  Farm        Farmers Market  Fast Food Restaurant   

     9th Most Common Venue 10th Most Common Venue  
0  New American Restaurant                Brewery  
1     Gym / Fitness Center    Japanese Restaurant  
2                     Café     Chinese Restaurant  
3     Fast Food Restaurant    Filipino Restaurant  
4      Filipino Restaurant      Fish & Chips Shop

In [39]:
venues_sorted.shape

(137, 11)

In [40]:
london_grouped.shape

(137, 336)

In [41]:
london_grouped=df

After our inspection of venues/facilities/amenities nearby the most profitable real estate investments in London, we could begin by clustering properties by venues/facilities/amenities nearby.

In [42]:
#Distribute in 5 Clusters

# set number of clusters
kclusters = 5

london_grouped_clustering = london_grouped.drop('Street', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(london_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50]

array([4, 4, 3, 4, 1, 2, 0, 2, 3, 3, 0, 4, 4, 4, 0, 2, 2, 2, 0, 4, 3, 0,
       1, 1, 4, 0, 4, 3, 3, 2, 2, 2, 2, 2, 4, 4, 2, 3, 2, 2, 1, 3, 4, 1,
       3, 1, 3, 3, 3, 4], dtype=int32)

In [43]:
london_grouped_clustering=df
london_grouped_clustering.head()

Street     Avg_Price   Latitude   Longitude
192     ALBION SQUARE  2.450000e+06 -41.273758  173.289393
384       ANHALT ROAD  2.435000e+06  51.480326   -0.166761
399   ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103
413   APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670
685  AYLESTONE AVENUE  2.286667e+06  51.540916   -0.217874

In [44]:
london_grouped_clustering.shape

(148, 4)

In [45]:
df.shape

(148, 4)

In [46]:
london_grouped_clustering.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [47]:
df.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [48]:
# add clustering labels
london_grouped_clustering['Cluster Labels'] = kmeans.labels_

# merge london_grouped with london_data to add latitude/longitude for each neighborhood
london_grouped_clustering = london_grouped_clustering.join(venues_sorted.set_index('Street'), on='Street')

london_grouped_clustering.head(30) # check the last columns!

Street     Avg_Price   Latitude   Longitude  \
192         ALBION SQUARE  2.450000e+06 -41.273758  173.289393   
384           ANHALT ROAD  2.435000e+06  51.480326   -0.166761   
399       ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103   
413       APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670   
685      AYLESTONE AVENUE  2.286667e+06  51.540916   -0.217874   
838       BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457   
961        BEAUCLERC ROAD  2.480000e+06  51.499577   -0.229033   
1082      BELVEDERE DRIVE  2.340000e+06  34.966857  -81.992424   
1195    BICKENHALL STREET  2.208500e+06  51.521197   -0.158934   
1233    BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468   
1530       BRAMPTON GROVE  2.456875e+06  51.570365   -0.283394   
1609    BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431   
1772             BROOKWAY  2.400000e+06  45.432185 -122.802812   
1885         BURBAGE ROAD  2.445000e+06  51.450967   -0.095288   
1949            BURY WALK  2.492500e+06  52.145529   -0.423593   
2030      CALEDONIAN ROAD  2.349182e+06  51.548482   -0.118359   
2035      CALLCOTT STREET  2.375000e+06  51.508350   -0.198328   
2096    CAMPDEN HILL ROAD  2.379653e+06  51.506461   -0.198896   
2103         CAMPION ROAD  2.461000e+06  52.681351    0.965439   
2125        CANNING PLACE  2.425000e+06  51.499570   -0.184248   
2192        CARLISLE ROAD  2.200000e+06  42.549099  -71.416669   
2197      CARLTON GARDENS  2.483500e+06 -37.801943  144.971970   
2209        CARLYLE COURT  2.300000e+06  32.626437  -83.717601   
2371       CHALCOT SQUARE  2.286679e+06  51.541455   -0.155265   
2449         CHARLES LANE  2.414000e+06  40.237525  -83.349229   
2525     CHELSEA CRESCENT  2.495000e+06  34.524738  -85.448539   
2569  CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081   
2600         CHEYNE COURT  2.250000e+06  51.599677    0.525623   
2769     CLARENDON STREET  2.250000e+06 -19.274357  146.800404   
2859     CLONCURRY STREET  2.388333e+06  51.473763   -0.216244   

      Cluster Labels          1st Most Common Venue  \
192                4                           Café   
384                4                            Pub   
399                3             Italian Restaurant   
413                4                            Pub   
685                1                           Park   
838                2                     Restaurant   
961                0                            Pub   
1082               2                   Burger Joint   
1195               3                          Hotel   
1233               3                            Pub   
1530               0      Middle Eastern Restaurant   
1609               4           Gym / Fitness Center   
1772               4                           Park   
1885               4                           Café   
1949               0                    Supermarket   
2030               2                  Grocery Store   
2035               2                            Pub   
2096               2                            Pub   
2103               0                    Sports Club   
2125               4                          Hotel   
2192               3                     Playground   
2197               0             Italian Restaurant   
2209               1                            NaN   
2371               1                           Café   
2449               4             Mexican Restaurant   
2525               0                            NaN   
2569               4                         Garden   
2600               3                      Gastropub   
2769               3  Paper / Office Supplies Store   
2859               2                           Café   

           2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
192            Indian Restaurant                   Pub             Restaurant   
384                Grocery Store                Garden     Chinese Restaurant   
399               Cloth

In [49]:
# Create Map

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_grouped_clustering['Latitude'], london_grouped_clustering['Longitude'], london_grouped_clustering['Street'], london_grouped_clustering['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [50]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 0, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price      1st Most Common Venue 2nd Most Common Venue  \
961   2480000.0                        Pub           Coffee Shop   
1530  2456875.0  Middle Eastern Restaurant           Men's Store   
1949  2492500.0                Supermarket    English Restaurant   
2103  2461000.0                Sports Club          Soccer Field   
2197  2483500.0         Italian Restaurant          Dessert Shop   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
961                  Hotel       Thai Restaurant         Grocery Store   
1530                  Lake               Factory                  Farm   
1949                   Pub                  Park   American Restaurant   
2103              Windmill                   Spa               Factory   
2197                 Hotel           Pizza Place                  Café   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
961              Gastropub           Chaat Place    Falafel Restaurant   
1530        Farmers Market  Fast Food Restaurant   Filipino Restaurant   
1949                  Café        Hardware Store   Rental Car Location   
2103    Falafel Restaurant                  Farm        Farmers Market   
2197     Convenience Store                   Bar        Ice Cream Shop   

          9th Most Common Venue 10th Most Common Venue  
961                 Supermarket           Cocktail Bar  
1530          Fish & Chips Shop            Fish Market  
1949                        Gym            Yoga Studio  
2103       Fast Food Restaurant    Filipino Restaurant  
2197  Middle Eastern Restaurant            Coffee Shop

In [51]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 1, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price       1st Most Common Venue 2nd Most Common Venue  \
685   2.286667e+06                        Park         Movie Theater   
2209  2.300000e+06                         NaN                   NaN   
2371  2.286679e+06                        Café                   Bar   
4221  2.265000e+06                         Pub           Flea Market   
4631  2.300000e+06  Construction & Landscaping           Yoga Studio   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
685               Bus Stop           Flea Market    Falafel Restaurant   
2209                   NaN                   NaN                   NaN   
2371           Pizza Place    Italian Restaurant                   Pub   
4221               Factory    Falafel Restaurant                  Farm   
4631           Flower Shop    Falafel Restaurant                  Farm   

     6th Most Common Venue          7th Most Common Venue  \
685                   Farm                 Farmers Market   
2209                   NaN                            NaN   
2371           Coffee Shop  Vegetarian / Vegan Restaurant   
4221        Farmers Market           Fast Food Restaurant   
4631        Farmers Market           Fast Food Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
685   Fast Food Restaurant   Filipino Restaurant      Fish & Chips Shop  
2209                   NaN                   NaN                    NaN  
2371            Restaurant          Cupcake Shop              Speakeasy  
4221   Filipino Restaurant     Fish & Chips Shop            Fish Market  
4631   Filipino Restaurant     Fish & Chips Shop            Fish Market

In [52]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 2, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
838   2.375000e+06            Restaurant                   Pub   
1082  2.340000e+06          Burger Joint        Discount Store   
2030  2.349182e+06         Grocery Store                  Café   
2035  2.375000e+06                   Pub                Bakery   
2096  2.379653e+06                   Pub                Bakery   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
838        Thai Restaurant           Pizza Place      Community Center   
1082              Friterie           Flower Shop                  Farm   
2030                  Park   Rental Car Location           Coffee Shop   
2035           Yoga Studio     Indian Restaurant                 Hotel   
2096           Coffee Shop                 Hotel         Grocery Store   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
838            Coffee Shop                  Park        Farmers Market   
1082        Farmers Market  Fast Food Restaurant   Filipino Restaurant   
2030           Yoga Studio               Theater      Greek Restaurant   
2035    Italian Restaurant         Grocery Store                  Park   
2096             Bookstore           Yoga Studio     Indian Restaurant   

     9th Most Common Venue 10th Most Common Venue  
838      Food & Drink Shop                   Café  
1082     Fish & Chips Shop            Fish Market  
2030                  Farm   Ethiopian Restaurant  
2035           Coffee Shop              Bookstore  
2096                Hostel         Ice Cream Shop

In [53]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 3, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue       2nd Most Common Venue  \
399   2250000.0    Italian Restaurant              Clothing Store   
1195  2208500.0                 Hotel                        Café   
1233  2217000.0                   Pub          Chinese Restaurant   
2192  2200000.0            Playground              Nightlife Spot   
2600  2250000.0             Gastropub  Construction & Landscaping   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
399                  Hotel                   Pub     Indian Restaurant   
1195    Chinese Restaurant            Restaurant             Gastropub   
1233         Train Station     French Restaurant                Bakery   
2192          Soccer Field           Yoga Studio           Fish Market   
2600           Yoga Studio           Flea Market    Falafel Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
399                 Bakery                Garden             Juice Bar   
1195     Indian Restaurant           Coffee Shop    Italian Restaurant   
1233               Brewery        Breakfast Spot           Coffee Shop   
2192    Falafel Restaurant                  Farm        Farmers Market   
2600                  Farm        Farmers Market  Fast Food Restaurant   

     9th Most Common Venue 10th Most Common Venue  
399                   Café     Chinese Restaurant  
1195                   Pub                 Garden  
1233                  Lake                   Farm  
2192  Fast Food Restaurant    Filipino Restaurant  
2600   Filipino Restaurant      Fish & Chips Shop

In [54]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 4, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue    2nd Most Common Venue  \
192   2450000.0                  Café        Indian Restaurant   
384   2435000.0                   Pub            Grocery Store   
413   2400000.0                   Pub                      Bar   
1609  2397132.0  Gym / Fitness Center  Health & Beauty Service   
1772  2400000.0                  Park              Flea Market   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
192                    Pub            Restaurant           Coffee Shop   
384                 Garden    Chinese Restaurant    English Restaurant   
413              Nightclub                Casino           Fish Market   
1609         Grocery Store     Indian Restaurant           Coffee Shop   
1772               Factory    Falafel Restaurant                  Farm   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
192                    Bar                Museum          Burger Joint   
384           Cocktail Bar                 Diner           Pizza Place   
413     Falafel Restaurant                  Farm        Farmers Market   
1609           Yoga Studio     Fish & Chips Shop    Falafel Restaurant   
1772        Farmers Market  Fast Food Restaurant   Filipino Restaurant   

        9th Most Common Venue 10th Most Common Venue  
192   New American Restaurant                Brewery  
384      Gym / Fitness Center    Japanese Restaurant  
413      Fast Food Restaurant    Filipino Restaurant  
1609                     Farm         Farmers Market  
1772        Fish & Chips Shop            Fish Market

# Results and Discussion section

We may discuss our results under two main perspectives.

First, we may examine them according to neighborhoods/London areas. It is interesting to note that, although West London (Notting Hill, Kensington, Chelsea, Marylebone) and North-West London (Hampsted) might be considered highly profitable venues to purchase a real estate according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores, South-West London (Wandsworth, Balham) and North-West London (Isliington) are arising as next future elite venues with a wide range of amenities and facilities. Accordingly, one might target under-priced real estates in these areas of London in order to make a business affair.

Second, we may analyze our results according to the five clusters we have produced. Even though, all clusters could praise an optimal range of facilities and amenities, we have found two main patterns. The first pattern we are referring to, i.e. Clusters 0, 2 and 4, may target home buyers prone to live in 'green' areas with parks, waterfronts. Instead, the second pattern we are referring to, i.e. Clusters 1 and 3, may target individuals who love pubs, theatres and soccer.

# Conclusion

To solve this above described business problem, we clustered London neighborhoods in order to recommend venues and the current average price of real estate where homebuyers can make a real estate investment. We recommended profitable venues according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores.

First, we gathered data on London properties and the relative price paid data were extracted from the HM Land Registry (http://landregistry.data.gov.uk/). Moreover, to explore and target recommended locations across different venues according to the presence of amenities and essential facilities, we accessed data through FourSquare API interface and arranged them as a data frame for visualization. By merging data on London properties and the relative price paid data from the HM Land Registry and data on amenities and essential facilities surrounding such properties from FourSquare API interface, we were able to recommend profitable real estate investments.

Second, The Methodology section comprised four stages: 1. Collect Inspection Data; 2. Explore and Understand Data; 3. Data preparation and preprocessing; 4. Modeling. In particular, in the modeling section, we used the k-means clustering technique as it is fast and efficient in terms of computational cost, is highly flexible to account for mutations in real estate market in London and is accurate.

Finally, we drew the conclusion that even though the London Housing Market may be in a rut, it is still an "ever-green" for business affairs. We discussed our results under two main perspectives. First, we examined them according to neighborhoods/London areas. although West London (Notting Hill, Kensington, Chelsea, Marylebone) and North-West London (Hampsted) might be considered highly profitable venues to purchase a real estate according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores, South-West London (Wandsworth, Balham) and North-West London (Isliington) are arising as next future elite venues with a wide range of amenities and facilities. Accordingly, one might target under-priced real estates in these areas of London in order to make a business affair. Second, we analyzed our results according to the five clusters we produced. While Clusters 0, 2 and 4 may target home buyers prone to live in 'green' areas with parks, waterfronts, Clusters 1 and 3 may target individuals who love pubs, theatres and soccer.